### PR TITLE
db and models setup

### PR DESCRIPTION
**Pull request by lugenx**
- connected to the database
- created mongoose models
- added routes
- created controller functions

some functions like user login and signup will be changed/improved when auth is applied.

### CODE REVIEW

File: server/src/controller/userController.js
Comment (mannan0819) [inline server/src/controller/userController.js:19]: We prob want to hash the users password and not want to store the exact string. 
I think possible Argon2 is whats good to use right now
https://www.npmjs.com/package/argon2

**Review by mannan0819 — COMMENTED** [review]
Just one comment to hash users passwords

File: server/src/controller/userController.js
Comment (lugenx) [reply server/src/controller/userController.js:19]: yeah, agree. we will hash it with argon2 or bcrypt, it will be encrypted together with salt.  

i think we can work on encryption when we work on user auth, which is another issue i have created. just tried to reduce the size of prs and focus on one at a time. but that's really good point and it's so critical we shouldn't miss.

how does the database structure look like to you guys, i tried but still i feel like maybe there should be more efficient way to do it? 🤔 . 

we will probably change it anyway while we progress but i would like to try to make it as close as possible.

let me know if you guys have any questions or suggestions while examine this.

**Comment by lugenx** [comment]
yeah, agree. we will hash it with argon2 or bcrypt, it will be encrypted together with salt.

i think we can work on encryption when we work on user auth, which is another issue i have created. just tried to reduce the size of prs and focus on one at a time. but that's really good point and it's so critical we shouldn't miss.

how does the database structure look like to you guys, i tried but still i feel like maybe there should be more efficient way to do it? 🤔 .

we will probably change it anyway while we progress but i would like to try to make it as close as possible.

let me know if you guys have any questions or suggestions while examine this.

**Comment by GrantMLambert** [comment]
We should also consider generating a random unique id string to each user and use that to identify them throughout the application as many apps do, as that will allow us to deal with situations in the future where it may be necessary to change a user's name, email or other information.

**Comment by lugenx** [comment]
per your guys' feedback, i updated the structure of the models, and added some additional fields.

There are three models:
- User
- Habit (questions)
- Answer: includes references of habit id and user id.

Here are some examples:

User:
image

Habit
image

Answer:
image

**Comment by lugenx** [comment]
> We should also consider generating a random unique id string to each user and use that to identify them throughout the application as many apps do, as that will allow us to deal with situations in the future where it may be necessary to change a user's name, email or other information.

so because the database assigns each user a unique id, i think it shouldn't be a problem to change user's name, email etc.

File: server/src/controller/userController.js
Comment (lugenx) [reply server/src/controller/userController.js:19]: also need to do some research about which one is better to choose, argon2 or Bcrypt.

Comment (AbhiBajpaishek) [reply server/src/controller/userController.js:19]: > We should also consider generating a random unique id string to each user and use that to identify them throughout the application as many apps do, as that will allow us to deal with situations in the future where it may be necessary to change a user's name, email or other information.

Mongo DB by default adds a unique id

**Comment by AbhiBajpaishek** [comment]
LGTM!!

**Review by AbhiBajpaishek — DISMISSED** [review]

**Comment by lugenx** [comment]
so i added the "roles" field to the user document in the mongoose schema in case we will have a different kinds of accounts in the future. see the example data below:

image

i will merge this pr after a review, rest improvements can be made later.

**Comment by lugenx** [comment]
@mannan0819 and @AbhiBajpaishek , i would like to see one of your approval before merging this. could you please review?

File: server/src/controller/userController.js
Comment (AbhiBajpaishek) [reply server/src/controller/userController.js:13]: @lugenx

**Review by AbhiBajpaishek — CHANGES_REQUESTED** [review]

File: server/src/controller/userController.js
Comment (lugenx) [reply server/src/controller/userController.js:13]: yeah, agree, much cleaner this way, let me update it.

Comment (lugenx) [reply server/src/controller/userController.js]: this is done! @AbhiBajpaishek

**Review by mannan0819 — APPROVED** [review]
Looks good to me. Just need to keep in mind to not store password strings.

**Pull request merged**
